extras/gizmo/axis-shapes.js

```diff
@@ -0,0 +1,650 @@
+    }
+
+    return colors;
+}
+
+class AxisShape {
+    _position;
+
+    _rotation;
+
+    _scale;
+
+    _layers = [];
+
+    _defaultColor;
+
+    _hoverColor;
+
+    device;
+
+    axis;
+
+    entity;
+
+    meshTriDataList = [];
+
+    meshInstances = [];
+
+    constructor(device, options) {
+        this.device = device;
+        this.axis = options.axis ?? 'x';
+        this._position = options.position ?? new Vec3();
+        this._rotation = options.rotation ?? new Vec3();
+        this._scale = options.scale ?? new Vec3(1, 1, 1);
+
+        this._layers = options.layers ?? this._layers;
+
+        this._defaultColor = options.defaultColor ?? Color.BLACK;
+        this._hoverColor = options.hoverColor ?? Color.WHITE;
+    }
+
+    hover(state) {
+        const material = state ? this._hoverColor : this._defaultColor;
+        for (let i = 0; i < this.meshInstances.length; i++) {
+            this.meshInstances[i].material = material;
+        }
+    }
+}
+
+class AxisArrow extends AxisShape {
+    _gap = 0;
+
+    _lineThickness = 0.04;
+
+    _lineLength = 0.5;
+
+    _arrowThickness = 0.15;
+
+    _arrowLength = 0.2;
+
+    _intersectTolerance = 0.1;
+
+    constructor(device, options = {}) {
+        super(device, options);
+
+        this.meshTriDataList = [
+            new MeshTriData(createCone(this.device)),
+            new MeshTriData(createCylinder(this.device))
+        ];
+
+        this._createArrow();
+    }
+
+    set gap(value) {
+        this._gap = value ?? 0;
+        this._updateHead();
+        this._updateLine();
+    }
+
+    get gap() {
+        return this._gap;
+    }
+
+    set lineThickness(value) {
+        this._lineThickness = value ?? 1;
+        this._updateHead();
+        this._updateLine();
+    }
+
+    get lineThickness() {
+        return this._lineThickness;
+    }
+
+    set lineLength(value) {
+        this._lineLength = value ?? 1;
+        this._updateHead();
+        this._updateLine();
+    }
+
+    get lineLength() {
+        return this._lineLength;
+    }
+
+    set arrowThickness(value) {
+        this._arrowThickness = value ?? 1;
+        this._updateHead();
+    }
+
+    get arrowThickness() {
+        return this._arrowThickness;
+    }
+
+    set arrowLength(value) {
+        this._arrowLength = value ?? 1;
+        this._updateHead();
+    }
+
+    get arrowLength() {
+        return this._arrowLength;
+    }
+
+    _createArrow() {
+        this.entity = new Entity('arrow:' + this.axis);
+        this.entity.setLocalPosition(this._position);
+        this.entity.setLocalEulerAngles(this._rotation);
+        this.entity.setLocalScale(this._scale);
+
+        // head
+        this._head = new Entity('head:' + this.axis);
+        this.entity.addChild(this._head);
+        this._updateHead();
+        let mesh = createShadowMesh(this.device, this._head, 'cone');
+        let meshInstance = new MeshInstance(mesh, this._defaultColor);
```

Comment:
Reminder: please make sure these meshInstances are destroyed at some point (or entity owning those using render component is destroyed)